File: src/ParticleEngine/ParticleEngine/Scripts/Particle/Physics/Materials/Liquid.cs
using static ParticleEngine.Particle.ParticleManager;

namespace ParticleEngine.Particle.Physics.Materials
{
    public static class Liquid
    {
        public static void Calculate(int row, int column, int rowSize, int columnSize, ref ParticleManager particleWorld)
        {
            ParticleInstance self = particleWorld.particles[row, column];

            if (row < rowSize - 1)
            {
                if (
                    (particleWorld.particles[row + 1, column].material == Material.Air ||
                    particleWorld.particles[row + 1, column].materialType == MaterialType.Gas) &&
                    particleWorld.particles[row, column].material == self.material)
                {
                    particleWorld.particles[row + 1, column] = self;
                    particleWorld.Clear(row, column);
                }
            }

            if (column > 0)
            {
                if (
                    (particleWorld.particles[row, column - 1].material == Material.Air ||
                    particleWorld.particles[row, column - 1].materialType == MaterialType.Gas) &&
                    particleWorld.particles[row, column].material == self.material)
                {
                    particleWorld.particles[row, column - 1] = self;
                    particleWorld.Clear(row, column);
                }
            }

            if (column < columnSize - 1)
            {
                if (
                    (particleWorld.particles[row, column + 1].material == Material.Air ||
                    particleWorld.particles[row, column + 1].materialType == MaterialType.Gas) &&
                    particleWorld.particles[row, column].material == self.material)
                {
                    particleWorld.particles[row, column + 1] = self;
                    particleWorld.Clear(row, column);
                }
            }
        }
    }
}

File: src/ParticleEnginePlusPlus/ParticleEngine/ParticleEngine/include/particle/physics/materials/liquid.cpp
#include <iostream>
#include "particle/particle_world.h"
#include "particle/particle_physics.h"

void calculate_liquid(int row, int col, ParticleWorld* particleWorld)
{
	ParticleWorld::ParticleInstance self = particleWorld->getParticle(row, col);

	if (particleWorld->canDown(row))
	{
		if (
			(particleWorld->getParticle(row + 1, col).material == ParticleWorld::Material::Air ||
				particleWorld->getParticle(row + 1, col).materialType == ParticleWorld::MaterialType::Gas ||
				particleWorld->getParticle(row + 1, col).material == ParticleWorld::Material::Fire) &&
			particleWorld->getParticle(row, col).material == self.material)
		{
			particleWorld->setParticle(row + 1, col, self);
			particleWorld->resetParticle(row, col);
		}
	}

	if (particleWorld->canLeft(col))
	{
		if (
			(particleWorld->getParticle(row, col - 1).material == ParticleWorld::Material::Air ||
				particleWorld->getParticle(row, col - 1).materialType == ParticleWorld::MaterialType::Gas ||
				particleWorld->getParticle(row, col - 1).material == ParticleWorld::Material::Fire) &&
			particleWorld->getParticle(row, col).material == self.material)
		{
			particleWorld->setParticle(row, col - 1, self);
			particleWorld->resetParticle(row, col);
		}
	}

	if (particleWorld->canRight(col))
	{
		if (
			(particleWorld->getParticle(row, col + 1).material == ParticleWorld::Material::Air ||
				particleWorld->getParticle(row, col + 1).materialType == ParticleWorld::MaterialType::Gas ||
				particleWorld->getParticle(row, col + 1).material == ParticleWorld::Material::Fire) &&
			particleWorld->getParticle(row, col).material == self.material)
		{
			particleWorld->setParticle(row, col + 1, self);
			particleWorld->resetParticle(row, col);
		}
	}
}